src/gluonts/model/seasonal_agg/_predictor.py

```diff
@@ -0,0 +1,133 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+from typing import Callable, Union
+
+import numpy as np
+
+from gluonts.core.component import validated
+from gluonts.dataset.common import DataEntry
+from gluonts.dataset.util import forecast_start
+from gluonts.dataset.field_names import FieldName
+from gluonts.model.forecast import Forecast, SampleForecast
+from gluonts.model.predictor import RepresentablePredictor
+from gluonts.transform.feature import (
+    LastValueImputation,
+    MissingValueImputation,
+)
+
+
+class SeasonalAggregatePredictor(RepresentablePredictor):
+    """
+    Seasonal aggegate forecaster.
+
+    For each time series :math:`y`, this predictor produces a forecast
+    :math:`\\tilde{y}(T+k) = f\big(y(T+k-h), y(T+k-2h), ...,
+    y(T+k-mh)\big)`, where :math:`T` is the forecast time,
+    :math:`k = 0, ...,` `prediction_length - 1`, :math:`m =`num_seasons`,
+    :math:`h =`season_length` and :math:`f =`agg_fun`.
+
+    If `prediction_length > season_length` :math:\times `num_seasons`, then the
+    seasonal aggregate is repeated multiple times. If a time series is shorter
+    than season_length` :math:\times `num_seasons`, then the `agg_fun` is
+    applied to the full time series.
+
+    Parameters
+    ----------
+    prediction_length
+        Number of time points to predict.
+    season_length
+        Seasonality used to make predictions. If this is an integer, then a
+        fixed sesasonlity is applied; if this is a function, then it will be
+        called on each given entry's ``freq`` attribute of the ``"start"``
+        field, and the returned seasonality will be used.
+    num_seasons
+        Number of seasons to aggregate.
+    agg_fun
+        Aggregate function.
+    imputation_method
+        The imputation method to use in case of missing values.
+        Defaults to :py:class:`LastValueImputation` which replaces each missing
+        value with the last value that was not missing.
+    """
+
+    @validated()
+    def __init__(
+        self,
+        prediction_length: int,
+        season_length: Union[int, Callable],
+        num_seasons: int,
+        agg_fun: Callable = np.nanmean,
```

Comment:
Just an idea: I was thinking that do we need `agg_fun` here? Unlike `SeasonalNaive` where we have one sample values per prediction step, here we have a list of values coming from different `num_seasons` for each prediction step. This can be thought of as prediction samples. Since we are using `SampleForecast` object as the return type, can we simply apply these aggregations like `mean`, `median (P50)`, `min (P1)`, and `max (P100)` directly on the `SampleForecast` object.

Reply:
I thought about it, but then the actual predictions of the model depends on some post-processing of `SampleForecast`, which might require some tinkering if we are evaluating it along with others. For example if we want to include the seasonal average method in some evaluation pipeline that accepts `Forecast` object and returns evaluation metrics, we need to tweak the output of `predict` method before evaluation otherwise the evaluation does not make sense.

I think it is better to have a simple contract for the models here: `predict` returns a `Forecast` object which in this case is a point forecast. We are using `SampleForecast` to represent point forecast is a different issue that we may need to address.